core/templates/pages/exploration-player-page/services/question-player-engine.service.spec.ts

```diff
@@ -0,0 +1,1099 @@
+    spyOn(contextService, 'isInQuestionPlayerMode').and.returnValue(true);
+    spyOn(urlService, 'getExplorationVersionFromUrl').and.returnValue(1);
+    spyOn(readOnlyExplorationBackendApiService, 'loadExplorationAsync')
+      .and.returnValue(Promise.resolve(exploration));
+
+    questionPlayerEngineService.init(
+      multipleQuestionsBackendDict, successHandler, failHandler);
+    let previewMode = questionPlayerEngineService.isInPreviewMode();
+
+    expect(previewMode).toBe(false);
+  });
+
+  it('should show warning message while loading a question ' +
+    'when question html content is null', () => {
+    let successHandler = jasmine.createSpy('success');
+    let failHandler = jasmine.createSpy('fail');
+
+    singleQuestionBackendDict.question_state_data
+      .content.html = null;
+    let alertsServiceSpy =
+      spyOn(alertsService, 'addWarning').and.returnValue();
+    let expressionInterpolationServiceSpy =
+      spyOn(expressionInterpolationService, 'processHtml')
+        .and.callFake((html, envs) => {
+          return html;
+        });
+
+    questionPlayerEngineService.init(
+      [singleQuestionBackendDict], successHandler, failHandler);
+
+    expect(alertsServiceSpy).toHaveBeenCalledWith('Expression parsing error.');
+    expect(expressionInterpolationServiceSpy).toHaveBeenCalled();
+  });
+
+  it('should call error callback if there are no questions', () => {
+    let successHandler = jasmine.createSpy('success');
+    let failHandler = jasmine.createSpy('fail');
+
+    questionPlayerEngineService.init(
+      [], successHandler, failHandler);
+    expect(successHandler).not.toHaveBeenCalled();
+    expect(failHandler).toHaveBeenCalled();
+  });
+
+  describe('on submiting answer ', () => {
+    it('should submit answer correctly', () => {
+      let successCallback = jasmine.createSpy('success');
+      let successHandler = jasmine.createSpy('success');
+      let failHandler = jasmine.createSpy('fail');
+      let answer = 'answer';
+      let interactionRulesService: InteractionRulesService;
+      let answerClassificationResult = new AnswerClassificationResult(
+        outcomeObjectFactory
+          .createNew('default', '', '', []), 1, 0, 'default_outcome'
+      );
+      answerClassificationResult.outcome.labelledAsCorrect = true;
+
+      let answerClassificationServiceSpy =
+        spyOn(answerClassificationService, 'getMatchingClassificationResult')
+          .and.returnValue(answerClassificationResult);
+
+      questionPlayerEngineService.init(
+        multipleQuestionsBackendDict, successHandler, failHandler);
+      questionPlayerEngineService.submitAnswer(
+        answer, interactionRulesService, successCallback);
+
+      expect(answerClassificationServiceSpy).toHaveBeenCalled();
+      expect(successCallback).toHaveBeenCalled();
+    });
+
+    it('should return when the answer is still being processed', () => {
```

Comment:
Don't understand the description -- return what? (Or to where?)

Try to write these descriptions similarly clearly to the ones in the portion of the code that has been reviewed.

Reply:
changed.